taichi/ui/backends/vulkan/app_context.cpp

```diff
@@ -0,0 +1,195 @@
+#include "taichi/ui/utils/utils.h"
+#include "taichi/ui/backends/vulkan/vulkan_utils.h"
+#include "taichi/ui/backends/vulkan/app_context.h"
+#include "taichi/ui/backends/vulkan/swap_chain.h"
+
+TI_UI_NAMESPACE_BEGIN
+
+namespace vulkan {
+
+using namespace taichi::lang::vulkan;
+
+const std::vector<const char *> device_extensions = {
+    VK_KHR_SWAPCHAIN_EXTENSION_NAME,
+    VK_KHR_EXTERNAL_MEMORY_EXTENSION_NAME,
+    VK_KHR_EXTERNAL_SEMAPHORE_EXTENSION_NAME,
+#ifdef _WIN64
+    VK_KHR_EXTERNAL_MEMORY_WIN32_EXTENSION_NAME,
+    VK_KHR_EXTERNAL_SEMAPHORE_WIN32_EXTENSION_NAME,
+#else
+    VK_KHR_EXTERNAL_MEMORY_FD_EXTENSION_NAME,
+    VK_KHR_EXTERNAL_SEMAPHORE_FD_EXTENSION_NAME,
+#endif
+};
+
+std::vector<const char *> get_required_instance_extensions() {
+  uint32_t glfw_ext_count = 0;
+  const char **glfw_extensions;
+  glfw_extensions = glfwGetRequiredInstanceExtensions(&glfw_ext_count);
+
+  std::vector<const char *> extensions(glfw_extensions,
+                                       glfw_extensions + glfw_ext_count);
+
+  extensions.push_back(VK_KHR_GET_PHYSICAL_DEVICE_PROPERTIES_2_EXTENSION_NAME);
```

Comment:
Always query support before adding an extension. Check the handling on KHR_surface extension

Reply:
Since technically a Vulkan backend device don't need to support these functionalities, it just means GGUI or Cuda->Vulkan GGUI is not supported

Reply:
Good point. Although this `std::vector<const char *>` is passed to `EmbeddedVulkanDevice` (in a previous PR), where the extensions are actually added. I will update the logic there to check the extensions before actually adding.

Reply:
Also remeber to add fields to VulkanCapabilities to tell the rest of the backend some features are enabled